kubernetes.go

```diff
@@ -0,0 +1,682 @@
+		}
+
+	}
+
+	data := map[string][]byte{"index.txt": []byte(indexTxt)}
+
+	if res, _ := openVPNPKI.secretCheckExists(secretIndexTxt); !res {
+		err = openVPNPKI.secretCreate(metav1.ObjectMeta{Name: secretIndexTxt}, data)
+	} else {
+		err = openVPNPKI.secretUpdate(metav1.ObjectMeta{Name: secretIndexTxt}, data)
+	}
+
+	return
+}
+
+func (openVPNPKI *OpenVPNPKI) updateIndexTxtOnDisk() (err error) {
+	secret, err := openVPNPKI.secretGetByName(secretIndexTxt)
+	indexTxt := secret.Data["index.txt"]
+	err = ioutil.WriteFile(fmt.Sprintf("%s/pki/index.txt", *easyrsaDirPath), indexTxt, 0600)
+	return
+}
+
+func (openVPNPKI *OpenVPNPKI) easyrsaGenCRL() (err error) {
+	err = openVPNPKI.indexTxtUpdate()
+	if err != nil {
+		return
+	}
+
+	secrets, err := openVPNPKI.secretsGetByLabels("index.txt=,type=clientAuth")
+	if err != nil {
+		return
+	}
+
+	var revoked []*RevokedCert
+
+	for _, secret := range secrets.Items {
+		if secret.Annotations["revokedAt"] != "" {
+			revokedAt, err := time.Parse(indexTxtDateFormat, secret.Annotations["revokedAt"])
+			if err != nil {
+				log.Warning(err)
+			}
+			cert, err := decodeCert(secret.Data[certFileName])
+			revoked = append(revoked, &RevokedCert{RevokedTime: revokedAt, Cert: cert})
+		}
+	}
+
+	crl, err := genCRL(revoked, openVPNPKI.CACert, openVPNPKI.CAPrivKeyRSA)
+	if err != nil {
+		return
+	}
+
+	secretData := map[string][]byte{
+		"crl.pem": crl.Bytes(),
+	}
+
+	err = openVPNPKI.secretCreate(metav1.ObjectMeta{Name: secretCRL}, secretData)
+
+	if res, _ := openVPNPKI.secretCheckExists(secretCRL); !res {
+		err = openVPNPKI.secretCreate(metav1.ObjectMeta{Name: secretCRL}, secretData)
+	} else {
+		err = openVPNPKI.secretUpdate(metav1.ObjectMeta{Name: secretCRL}, secretData)
+	}
+
+	return
+}
+
+func (openVPNPKI *OpenVPNPKI) easyrsaBuildClient(commonName string) (err error) {
+	// check certificate exists
+	_, err = openVPNPKI.secretGetByLabels("name=" + commonName)
+	if err == nil {
+		return errors.New(fmt.Sprintf("certificate for user (%s) already exists", commonName))
+	}
+
+	clientPrivKeyPEM, err := genPrivKey()
+	if err != nil {
+		return
+	}
+
+	clientPrivKeyRSA, err := decodePrivKey(clientPrivKeyPEM.Bytes())
+	if err != nil {
+		return
+	}
+
+	clientCertPEM, _ := genClientCert(clientPrivKeyRSA, openVPNPKI.CAPrivKeyRSA, openVPNPKI.CACert, commonName)
+	clientCert, err := decodeCert(clientCertPEM.Bytes())
+
+	secretData := map[string][]byte{
+		certFileName:    clientCertPEM.Bytes(),
+		privKeyFileName: clientPrivKeyPEM.Bytes(),
+	}
+
+	name := fmt.Sprintf(secretClientTmpl, clientCert.SerialNumber)
+	annotations := map[string]string{
+		"commonName":   commonName,
+		"notBefore":    clientCert.NotBefore.Format(indexTxtDateFormat),
+		"notAfter":     clientCert.NotAfter.Format(indexTxtDateFormat),
+		"revokedAt":    "",
+		"serialNumber": clientCert.SerialNumber.String(),
+	}
+	labels := map[string]string{
+		"index.txt": "",
+		"type":      "clientAuth",
+		"name":      commonName,
+	}
```

Comment:
It would be nice to add some understandable global label for all those certificates, eg something like:

```
app.kubernetes.io/managed-by: ovpn-admin
```

Take a look at Kubernetes recomended labels documentation page
https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/